### PR TITLE
Deployment detail backend

### PR DESCRIPTION
**Pull request by Draiken**
Still WIP

Not sure how to deal with the `NewReplicaSet` and `OldReplicaSets` fields. Are they really needed for the frontend?

Part of #595

---
This change is [Reviewable](https://reviewable.io/reviews/kubernetes/dashboard/750)

### CODE REVIEW

**Comment by Draiken** [comment]
I signed it!

**Comment by bryk** [comment]
@floreks @digitalfishpond Can you review this?

**Comment by bryk** [comment]
Hey, this is nice! Can you make it merge-able so that we merge it and do more work in follow up PRs? We like small PRs here :) 

> Not sure how to deal with the NewReplicaSet and OldReplicaSets fields. Are they really needed for the frontend?

I think it is worth to show them if we can explain easily to users what they are.



---

Reviewed 1 of 3 files at r1, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-:-KICgl4OyXo-CFFMBJTV:1366942502)*

**Comment by floreks** [comment]
Old/new replica sets are somehow related to deployment so we should show them. One approach could be to return `ReplicaSetList` for both from backend and display them using `replicasetcardlist` component with appropriate title.

@bryk what do you think?

This PR looks really nice! Just few comments.

<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> @floreks @digitalfishpond Can you review this? 
</details>


---

Reviewed 1 of 3 files at r1, 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[src/app/backend/resource/deployment/deploymentdetail.go, line 26 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-KICgEtG_uTUhSszKcxW:-KICgEtHqnctOb-KXtIH:-1206284560) ([raw file](https://github.com/kubernetes/dashboard/blob/dfc5e16a856949ce566de3b88affd42f6a07da58/src/app/backend/resource/deployment/deploymentdetail.go#L26)):*
> ```Go
> 
> 	// Status
> 	Status extensions.DeploymentStatus `json:"status"`
> ```

Please use our structures. For this you could add structure similar to `PodInfo` with status info.

---

*[src/app/backend/resource/deployment/deploymentdetail.go, line 38 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-KICgM25eI7I87RdHUN9:-KICgM25eI7I87RdHUNA:842284457) ([raw file](https://github.com/kubernetes/dashboard/blob/dfc5e16a856949ce566de3b88affd42f6a07da58/src/app/backend/resource/deployment/deploymentdetail.go#L38)):*
> ```Go
> 
> 	// OldReplicaSets
> 	OldReplicaSets []extensions.ReplicaSet `json:"oldReplicaSets"`
> ```

This can be our `ReplicaSet` object.

---

*[src/app/backend/resource/deployment/deploymentdetail.go, line 41 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-KICgXi3bgmfimdLeGC5:-KICgXi3bgmfimdLeGC6:1854746597) ([raw file](https://github.com/kubernetes/dashboard/blob/dfc5e16a856949ce566de3b88affd42f6a07da58/src/app/backend/resource/deployment/deploymentdetail.go#L41)):*
> ```Go
> 
> 	// New replica set used by this deployment
> 	NewReplicaSet extensions.ReplicaSet `json:"newReplicaSet"`
> ```

Same here

---

*[src/app/backend/resource/deployment/deploymentdetail.go, line 89 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-KICgrbJzaC8NbzbVItf:-KICgrbKjs4DW0B8Kt_l:-203505725) ([raw file](https://github.com/kubernetes/dashboard/blob/dfc5e16a856949ce566de3b88affd42f6a07da58/src/app/backend/resource/deployment/deploymentdetail.go#L89)):*
> ```Go
> }
> 
> func getDeploymentDetail(deployment *extensions.Deployment, old []*extensions.ReplicaSet, newReplicaSet *extensions.ReplicaSet, events *common.EventList) *DeploymentDetail {
> ```

Could you use gofmt to format the code? Some lines are getting pretty long.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-:-KICetkIvk1ptpVHJAg0:654632173)*

**Comment by bryk** [comment]
And yeah, only minor comments. IMO, we can merge this soon.  

<details><summary><i>Previously, floreks (Sebastian Florek) wrote…</i></summary>
> Old/new replica sets are somehow related to deployment so we should show them. One approach could be to return `ReplicaSetList` for both from backend and display them using `replicasetcardlist` component with appropriate title.
> 
> @bryk what do you think?
> 
> This PR looks really nice! Just few comments.
</details>


---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-:-KICh_ajYvbBOQCKD0PD:2141580844)*

File: src/app/backend/resource/deployment/deploymentdetail.go [hunk 1]
```diff
@@ -104,8 +116,33 @@ func getDeploymentDetail(deployment *extensions.Deployment, old []*extensions.Re
...
-		OldReplicaSets: oldReplicaSets,
-		NewReplicaSet:  *newReplicaSet,
-		EventList:      *events,
+		OldReplicaSetList: oldReplicaSetList,
+		NewReplicaSet:     newReplicaSet,
+		EventList:         *events,
+	}
+}
+
+func toReplicaSetList(resourceList []extensions.ReplicaSet, pods []api.Pod) replicaset.ReplicaSetList {
```
Comment (Draiken) [inline src/app/backend/resource/deployment/deploymentdetail.go]: There's a bit of duplication here on transforming extension replicaset objects into the dashboard ones. I'm not very familiar with Go or the source code, so I wasn't sure if I should, for example, export the getReplicaSetList on the replicaset package.

**Comment by Draiken** [comment]
I'm using gofmt and it didn't change the line sizes. Sorry I didn't know I had to do it manually.

**Comment by bryk** [comment]
Last cosmetic comment and I think we can merge :) @floreks Also take a look, if you can. 

<details><summary><i>Previously, Draiken (Luiz Felipe G. Pereira) wrote…</i></summary>
> I'm using gofmt and it didn't change the line sizes. Sorry I didn't know I had to do it manually.
</details>


---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/app/backend/resource/deployment/deploymentdetail.go, line 125 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-KI9gmnHjUt7oXNiiJJQ-r4-125:-KIRfxJsCtmFeeRnEFc5:-1936803685) ([raw file](https://github.com/kubernetes/dashboard/blob/93c0cb0c7f7c221663f84814ef1ded8d7c39e1cc/src/app/backend/resource/deployment/deploymentdetail.go#L125)):*
> ... so I wasn't sure if I should, for example, export the getReplicaSetList on the replicaset package.

Yeah, go ahead and export it here. In general, try to avoid copy & paste when possible. 

To export getReplicaSetList just start it with Capital Letter. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-:-KIRgWQg1P1i9a2PflEN:-1852582390)*

**Comment by bryk** [comment]
:lgtm: Merging. If you need any more help, talk to me or us on slack: https://kubernetes.slack.com/messages/sig-ui/

<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Last cosmetic comment and I think we can merge :) @floreks Also take a look, if you can. 
</details>


---

Reviewed 1 of 3 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/750#-:-KIWS8dcP5MVX3Xs5CP7:-22148706)*

**Pull request merged**